OData/src/System.Web.OData/OData/Query/OrderByQueryOption.cs

```diff
@@ -245,6 +236,25 @@ private IOrderedQueryable ApplyToCore(IQueryable query, ODataQuerySettings query
                     }
                     alreadyOrdered = true;
                 }
+                else if (openPropertyNode != null)
+                {
+                    // This check prevents queries with duplicate properties (e.g. $orderby=Id,Id,Id,Id...) from causing stack overflows
+                    if (propertiesSoFar.Contains(openPropertyNode.PropertyName))
+                    {
+                        throw new ODataException(Error.Format(SRResources.OrderByDuplicateProperty, openPropertyNode.PropertyName));
+                    }
+                    propertiesSoFar.Add(openPropertyNode.PropertyName);
+
+                    if (openPropertyNode.OrderByClause != null)
+                    {
+                        querySoFar = AddOrderByQueryForProperty(query, querySettings, openPropertyNode.OrderByClause, querySoFar, openPropertyNode.Direction, alreadyOrdered);
+                    }
+                    else
+                    {
+                        throw new NotImplementedException(string.Format("Error ordering the dynamic property {0} on type {1}", openPropertyNode.PropertyName, Context.ElementClrType.FullName));
```

Comment:
Please add it into resource string, which will be translate into multiple language